### PR TITLE
Build against Java 16

### PR DESCRIPTION
**Pull request by f4lco**
This was a lot more intricate than I thought. This PR solves two main issues:
- To support Java 16, we need to have a Gradle wrapper that also supports it: 7.0-rc1.
- I removed the Gradle wrapper task definitions from integration tests. The CI build uses the Gradle wrapper in the root directory anyways. If you see any use in the prior task definitons, please object.
- Java 16 adds strict validation for modules; however, because this causes only errors in project using Spring, I fixed this within the integration test projects. I do not think there is a one-size-fits-all general solution to this.
- I know that the `JavaVersion.current().isJava9Compatible()` switch is not strictly required for master, because master does not build against Java 8. However, I think that there is some value to using this patch as-is in 3.x. I'm not sure if those branches should drift further apart than ultimately necessary.

### CODE REVIEW

**Comment by boris-petrov** [comment]
Great! That's exactly the solution I was going to propose. I saw you trying to first install the wrapper with Java 11 and then to run the tests with the specific version and I was wondering why you just don't update the wrapper version in the repo. And that's what you did here so that's great! Thanks!

I'll leave one single comment inline right now.

File: build.gradle [hunk 1]
```diff
@@ -106,7 +106,7 @@ task cleanIntegrationTests(type: GradleBuild) {
 }
 
 wrapper {
```
Comment (boris-petrov) [inline build.gradle:108]: Why is this `wrapper` needed here anyways? I see you've removed the other 2. Why is this needed?

**Comment by boris-petrov** [comment]
Also, why is the wrapper JAR also added to the repo? Was wasn't it there before? Is that something new in Gradle 7?

**Comment by boris-petrov** [comment]
And thirdly - there is an RC 2 of Gradle 7, you might want to update it everywhere.

File: build.gradle (hunk 1 above)
Comment (f4lco) [reply build.gradle:108]: I see that the task options given here only repeat what's already in the GitHub actions file. I'll remove it, thanks.

**Comment by f4lco** [comment]
> Also, why is the wrapper JAR also added to the repo? Was wasn't it there before? Is that something new in Gradle 7?

Actually, this JAR file is already present in the repo. We need one binary which figures out how to download a Gradle distribution. It hasn't been updated in a while, last time in 798ddc369bb2378a726f40a58e7bc7670c819e96. It seems like we are still using a Gradle wrapper JAR of Gradle 4.9 to download Gradle 6 or 7 for us. According to [Gradle docs](https://docs.gradle.org/current/userguide/gradle_wrapper.html#sec:adding_wrapper) there's no way around putting the JAR in version control. It's easy to see:

```
$ rm gradle/wrapper/gradle-wrapper.jar
$ ./gradlew help
Error: Could not find or load main class org.gradle.wrapper.GradleWrapperMain
Caused by: java.lang.ClassNotFoundException: org.gradle.wrapper.GradleWrapperMain
```

> And thirdly - there is an RC 2 of Gradle 7, you might want to update it everywhere.

Yes, thank you. Will try.

**Comment by boris-petrov** [comment]
OK, this seems great now, merge at will! Thanks again. :)
